delete.js

```diff
@@ -181,83 +231,139 @@ async function getHarvestedFiles(uri) {
  * @param {string} submissionDocument URI of the submitted document to get the related file for
  * @param {string} fileType URI of the type of the related file
  */
-async function getTTLResource(submissionDocument, fileType) {
+async function getTTLResource(submissionDocument, fileType, graph) {
   const response = await querySudo(`
     PREFIX mu: <http://mu.semte.ch/vocabularies/core/>
     PREFIX dct: <http://purl.org/dc/terms/>
     PREFIX nie: <http://www.semanticdesktop.org/ontologies/2007/01/19/nie#>
 
-    SELECT DISTINCT ?file
+    SELECT DISTINCT ?logical ?physical
     WHERE {
-      GRAPH ?g {
-        ${sparqlEscapeUri(submissionDocument)} dct:source ?file .
-      }
-      GRAPH ${sparqlEscapeUri(FILE_GRAPH)} {
-        ?file dct:type ${sparqlEscapeUri(fileType)} .
+      GRAPH ${sparqlEscapeUri(graph)} {
+        ${sparqlEscapeUri(submissionDocument)} dct:source ?physical .
+        ?physical
+          dct:type ${sparqlEscapeUri(fileType)} ;
+          nie:dataSource ?logical .
       }
     } LIMIT 1
   `);
 
   const parser = new SparqlJsonParser();
   const parsedResults = parser.parseJsonResults(response);
-  if (parsedResults.length < 1)
+  if (parsedResults.length < 1) {
     console.log(
       `Part of type ${fileType} for submission document ${submissionDocument} not found`,
     );
-  return parsedResults[0]?.file?.value;
+    return;
+  }
+  return {
+    logical: parsedResults[0]?.logical?.value,
+    physical: parsedResults[0]?.physical?.value,
+  };
 }
 
 /**
  * Delete submission resource
  *
  * @param {string} URI of the resource to delete the related files for
  */
-async function deleteResource(URI) {
-  return querySudo(`
+async function deleteResource(uri, graph) {
+  return updateSudo(`
     DELETE {
-      GRAPH ?g {
-        ${sparqlEscapeUri(URI)} ?p ?o .
+      GRAPH ${sparqlEscapeUri(graph)} {
+        ${sparqlEscapeUri(uri)} ?p ?o .
       }
     }
     WHERE {
-      GRAPH ?g {
-        ${sparqlEscapeUri(URI)} ?p ?o .
+      GRAPH ${sparqlEscapeUri(graph)} {
+        ${sparqlEscapeUri(uri)} ?p ?o .
       }
     }
   `);
 }
 
-async function getSubmissionById(submissionId) {
+async function deleteTaskwithJob(taskUri, graph) {
+  return updateSudo(`
+    PREFIX dct: <http://purl.org/dc/terms/>
+    PREFIX task: <http://redpencil.data.gift/vocabularies/tasks/>
+
+    DELETE {
+      GRAPH ${sparqlEscapeUri(graph)} {
+        ?hc2 ?hp2 ?ho2 .
+        ?hc ?hp ?ho .
+        ?rc ?rp ?ro .
+        ?ic ?ip ?io .
+        ?task ?tp ?to .
+        ?job ?jp ?jo .
+      }
+    } WHERE {
+      GRAPH ${sparqlEscapeUri(graph)} {
+        ${sparqlEscapeUri(taskUri)} dct:isPartOf ?job .
+        ?job ?jp ?jo .
+        ?task
+          dct:isPartOf ?job ;
+          ?tp ?to .
+
+        OPTIONAL {
```

Comment:
Perhaps replacing the optional with a union, and the many option with values?
something along the lines

```
{
// the base query
}
UNION {
  VALUES ?predicate {
  task:resultsContainer
  ....
}
}
```

Reply:
Turns out that is not a good solution, but the original solution was also very bad! Because of the `OPTIONAL` expressions, some variables might be unbound and following `OPTIONAL` patterns that use those variables can then bind to anything else that could be unrelated to the Task and Job from before. The solution is to put everything in a single `OPTIONAL` block and to repeat the `OPTIONAL` block with less specific patterns afterwards as a fallback. This new solution (that has been tested a bit better) is already committed.